kairon/shared/utils.py

```diff
@@ -1785,17 +1812,23 @@ def validate_steps(steps: List):
         if recursive_simple_cycles(story_graph):
             raise AppException("Story cannot contain cycle!")
 
-        leaf_nodes_with_intent = [node for node in story_graph.nodes() if story_graph.out_degree(node) == 0
-                                    and node.step_type == 'INTENT']
+        leaf_nodes_with_intent = StoryValidator.get_leaf_nodes_with_intent(story_graph)
         if leaf_nodes_with_intent:
             raise AppException("Leaf nodes cannot be intent")
 
         for story_node in story_graph.nodes():
             if story_node.step_type == "INTENT":
                 if [successor for successor in story_graph.successors(story_node) if successor.step_type == "INTENT"]:
-                    raise AppException("Intent should be followed by an Action")
+                    raise AppException("Intent should be followed by an Action or Slot type event")
                 if len(list(story_graph.successors(story_node))) > 1:
-                    raise AppException("Intent can only have one connection of action type")
+                    raise AppException("Intent can only have one connection of action type or slot type")
+
+        leaf_nodes = StoryValidator.get_leaf_nodes(story_graph)
+        leaf_node_ids = [value.node_id for value in leaf_nodes]
+        if path:
+            for value in path:
+                if value['node_id'] not in leaf_node_ids:
+                    raise ValidationError("Only leaf nodes can be tagged with a flow")
```

Comment:
not worth iterating again just to validate slot events. try using the loop in validate steps line 1819.